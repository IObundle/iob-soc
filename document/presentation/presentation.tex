\documentclass [xcolor=svgnames, t] {beamer}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{booktabs, comment}
\usepackage[absolute, overlay]{textpos}
\useoutertheme{infolines}
\setbeamercolor{title in head/foot}{bg=internationalorange}
\setbeamercolor{author in head/foot}{bg=dodgerblue}
\usepackage{csquotes}
\usepackage[style=verbose-ibid,backend=bibtex]{biblatex}
\bibliography{bibfile}
\usepackage{amsmath}
\usepackage[makeroom]{cancel}
\usepackage{textpos}
\usepackage{tikz}
\usepackage{listings}
\graphicspath{ {./figures/} }
\usepackage{hyperref}
\hypersetup{
    colorlinks=true,
    linkcolor=blue,
    filecolor=magenta,
    urlcolor=cyan,
}

\usetheme{Madrid}
\definecolor{myuniversity}{RGB}{0, 60, 113}
\definecolor{internationalorange}{RGB}{231, 93,  42}
 	\definecolor{dodgerblue}{RGB}{0, 119,202}
\usecolortheme[named=myuniversity]{structure}

\title[IOb-SoC Presentation]{IOb-SoC}
\subtitle{Tutorial: Create a RISC-V-based System-on-Chip}
\institute[IObundle Lda.]{IObundle Lda.}
\titlegraphic{\includegraphics[height=2.5cm]{Logo.png}}
%\author[Jos√© T. de Sousa]{Jos\'e T. de Sousa}
%\institute[IObundle Lda]{IObundle Lda}
\date{\today}


\addtobeamertemplate{navigation symbols}{}{%
    \usebeamerfont{footline}%
    \usebeamercolor[fg]{footline}%
    \hspace{1em}%
    \insertframenumber/\inserttotalframenumber
}

\begin{document}

\begin{frame}
 \titlepage
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\logo{\includegraphics[scale=0.2]{Logo.png}~%
}
%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{frame}{Outline}
\begin{center}
   \begin{itemize}
     \item Introduction
     \item Project setup
     \item Instantiate an IP core in your SoC
     \item Write the software to drive the new peripheral
     \item Simulate IOb-SoC
     \item Run IOb-SoC on an FPGA board
     \item Conclusion
 \end{itemize}
\end{center}
\end{frame}


\begin{frame}{Introduction}
\begin{center}
    \begin{itemize}
      \item Building processor-based systems from scratch is challenging
      \item The IOb-SoC template eases this task by providing a base Verilog SoC equipped with
        \begin{itemize}
        \item a RISC-V CPU
        \item a memory system including boot ROM, RAM, 2-level cache system and an AXI4 interface to external memory (DDR)
        \item a UART communications module
        \item an example firmware program
        \end{itemize}
      \item Users can add IP cores and software to build their SoCs
      \item This tutorial exemplifies the addition of a Timer IP core and the use of its software driver
    \end{itemize}
\end{center}
\end{frame}

\begin{frame}{Project setup}
\begin{center}
  \begin{itemize}
    \item Use a Linux real or virtual machine (see the README file to download a VM)
    \item Make sure a {\bf stable} version of the open source Icarus Verilog simulator (\url{iverilog.icarus.com}) is installed locally or on some remote server
    \item Make sure you have FPGA build tools installed locally or on some remote server
    \item Make sure you have an FPGA board attached to your Linux machine or to some remote server
    \item Set up {\bf ssh} access key to GitHub (\url{github.com}) (using https will ask for your password many times)
    \item Follow the instructions in the IOb-SoC repository's README file to clone the repository and install the tools
  \end{itemize}
\end{center}
\end{frame}


\begin{frame}{Instantiate an IP core in your SoC}
  \begin{itemize}
  \item The Timer IP core at \url{github.com/IObundle/iob-timer.git} is used here as an example
  \item Add the Timer IP core repository as a git submodule of your IOb-SoC clone repository:\\
    {\tt \tiny git submodule add git@github.com:IObundle/iob-timer.git submodules/TIMER}
  \item Update the Timer IP core submodules:\\
    {\tt \tiny git submodule update ----init ----recursive}
  \item Add the Timer IP core to the list of peripherals in the {\tt ./config.mk} file:\\
    {\tt PERIPHERALS:=UART {\em TIMER}}
  \item Include the Timer IP core {\tt hardware.mk} file in {\tt ./hardware.mk} file:\\
    {\tt \tiny include \$(TIMER\_DIR)/hardware/hardware.mk}
  \item Include the Timer IP core {\tt embedded.mk} file in {\tt ./firmware/Makefile} file:\\
    {\tt \tiny include \$(TIMER\_DIR)/software/embedded/embedded.mk}
  \end{itemize}
\end{frame}

\begin{frame}{Instantiate an IP core in your SoC}
  \begin{itemize}
  \item Include the Timer IP core {\tt pc.mk} file in {\tt ./pc-emul/Makefile} file:\\
    {\tt \tiny include \$(TIMER\_DIR)/software/pc-emul/pc.mk}
  \item An IP core can be integrated into IOb-SoC if it provides the following files:
    \begin{itemize}
    \item CORE\_REPO/hardware/hardware.mk
    \item CORE\_REPO/software/embedded/embedded.mk
    \item CORE\_REPO/software/pc-emul/pc.mk
    \end{itemize}
  \item Study these files and its references in the Timer IP core repository.
  \end{itemize}
\end{frame}

\begin{frame}[fragile]{Write the software to drive the new peripheral}
  Edit the {\tt software/firmware/firmware.c} file to look as follows
  \begin{tiny}
    \begin{lstlisting}
#include "system.h"
#include "periphs.h"
#include "iob-uart.h"
#include "printf.h"

#include "iob_timer.h"

int main()
{
  unsigned long long elapsed;
  unsigned int elapsedu;

  //init timer and uart
  timer_init(TIMER_BASE);
  uart_init(UART_BASE, FREQ/BAUD);

  printf("\nHello world!\n");

  //read current timer count, compute elapsed time
  elapsed  = timer_get_count();
  elapsedu = timer_time_us();

  printf("\nExecution time: %d clock cycles\n", (unsigned int) elapsed);
  printf("\nExecution time: %dus @%dMHz\n\n", elapsedu, FREQ/1000000);

  uart_finish();
}
    \end{lstlisting}
  \end{tiny}
\end{frame}


\begin{frame}[fragile]{Simulate IOb-SoC}
\begin{itemize}
\item Run the simulation with the firmware pre-initialised in the memory:\\
  {\tt make sim}
\item The printed messages show that the firmware and bootloader C files are compiled
\item The whole system's Verilog files are also compiled
\item Finally the simulation is started and the following is printed:
\end{itemize}

\begin{tiny}
  \begin{lstlisting}

TESTBENCH: connecting..IOb-Bootloader: connected!
IOb-Bootloader: Restart CPU to run user program...

Hello world!

Execution time: 4356 clock cycles

Execution time: 44us @100MHz

TESTBENCH: exiting

  \end{lstlisting}
\end{tiny}
\end{frame}


\begin{frame}{Run IOb-SoC on an FPGA board (1)}
\begin{itemize}
\item To compile and run your SoC in one of our FPGA boards, contacts us at info@iobundle.com.
\item To compile and run your SoC on your FPGA board, add a directory into {\tt ./hardware/fpga/<toolchain>}, where <toolchain> is either *vivado* or *quartus*. Use the existing board directories as examples
\item Then issue the following command:\\
  {\tt make run BOARD=<board\_dir\_name> INIT\_MEM=0}\\
  This will compile the software and the hardware, produce an FPGA bitstream,
  load it to the device, load the firmware binary using the UART ({\tt INIT\_MEM=0} prevents the FPGA memory initialisation), start the
  program and direct the standard output to your PC terminal.
\item If you change only the firmware and repeat the above command, only the
  firmware will be recompiled, reloaded and rerun
\end{itemize}
\end{frame}


\begin{frame}[fragile]{Run IOb-SoC on an FPGA board (2)}
\begin{itemize}
\item When running IOb-SoC on an FPGA with the default settings and the firmware
  pre-initialised in the memory, the following should be printed:
\end{itemize}

\begin{tiny}
  \begin{lstlisting}

+-----------------------------------------------+
|                   IOb-Console                 |
+-----------------------------------------------+

  BaudRate = 115200
  StopBits = 1
  Parity   = None

IOb-Console: connecting...

IOb-Bootloader: connected!
IOb-Bootloader: Restart CPU to run user program...

Hello world!

Execution time: 114466 clock cycles

Execution time: 1145us @100MHz

IOb-Console: exiting...

  \end{lstlisting}
\end{tiny}
\end{frame}

\begin{frame}{Conclusion}
  \begin{itemize}
  \item A tutorial on creating a simple SoC using IOb-SoC has been presented
  \item The addition of an example peripheral IP core has been illustrated
  \item A simple firmware that uses the IP core driver functions has been explained
  \item RTL simulation of the system running the firmware has been demonstrated
  \item FPGA board running of the system running the firmware has been demonstrated
  \end{itemize}
\end{frame}

% for including figures in slides:
%\begin{frame}{Introduction}
%\begin{center}
%  \begin{columns}[onlytextwidth]
%    \column{0.5\textwidth}
%    bla
%    \column{0.5\textwidth}
%    \begin{figure}
%      \centering
%       \includegraphics[width=0.9\textwidth]{turb.jpg}
%      \caption{1. Flow visualisation (source: www.bronkhorst.com).}
%      \label{fig:my_label}
%    \end{figure}
%  \end{columns}
%\end{center}
%\end{frame}

\end{document}
