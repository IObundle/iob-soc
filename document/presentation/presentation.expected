\relax 
\providecommand\hyper@newdestlabel[2]{}
\providecommand\HyperFirstAtBeginDocument{\AtBeginDocument}
\HyperFirstAtBeginDocument{\ifx\hyper@anchor\@undefined
\global\let\oldcontentsline\contentsline
\gdef\contentsline#1#2#3#4{\oldcontentsline{#1}{#2}{#3}}
\global\let\oldnewlabel\newlabel
\gdef\newlabel#1#2{\newlabelxx{#1}#2}
\gdef\newlabelxx#1#2#3#4#5#6{\oldnewlabel{#1}{{#2}{#3}}}
\AtEndDocument{\ifx\hyper@anchor\@undefined
\let\contentsline\oldcontentsline
\let\newlabel\oldnewlabel
\fi}
\fi}
\global\let\hyper@last\relax 
\gdef\HyperFirstAtBeginDocument#1{#1}
\providecommand\HyField@AuxAddToFields[1]{}
\providecommand\HyField@AuxAddToCoFields[2]{}
\bibstyle{biblatex}
\bibdata{presentation-blx,bibfile}
\citation{biblatex-control}
\abx@aux@refcontext{nty/global//global/global}
\@writefile{toc}{\boolfalse {citerequest}\boolfalse {citetracker}\boolfalse {pagetracker}\boolfalse {backtracker}\relax }
\@writefile{lof}{\boolfalse {citerequest}\boolfalse {citetracker}\boolfalse {pagetracker}\boolfalse {backtracker}\relax }
\@writefile{lot}{\boolfalse {citerequest}\boolfalse {citetracker}\boolfalse {pagetracker}\boolfalse {backtracker}\relax }
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{1}{1/1}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {1}{1}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{2}{2/2}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {2}{2}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{3}{3/3}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {3}{3}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{4}{4/4}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {4}{4}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{5}{5/5}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {5}{5}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{6}{6/6}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {6}{6}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{7}{7/7}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {7}{7}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{8}{8/8}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {8}{8}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{9}{9/9}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {9}{9}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{10}{10/10}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {10}{10}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\slideentry {0}{0}{11}{11/11}{}{0}}}
\@writefile{nav}{\defcounter {refsection}{0}\relax }\@writefile{nav}{\headcommand {\beamer@framepages {11}{11}}}
\@writefile{nav}{\headcommand {\beamer@partpages {1}{11}}}
\@writefile{nav}{\headcommand {\beamer@subsectionpages {1}{11}}}
\@writefile{nav}{\headcommand {\beamer@sectionpages {1}{11}}}
\@writefile{nav}{\headcommand {\beamer@documentpages {11}}}
\@writefile{nav}{\headcommand {\gdef \inserttotalframenumber {11}}}
